### ISSUE TITLE
Logistic regression - features with zero weight associated with non-zero SHAP values

### ISSUE THREAD

**Issue by inezpereira** [shap/shap#3170]
Creating an issue in case this is not the expected behavior and should be documented as a bug.

### Discussed in https://github.com/shap/shap/discussions/3160

<div type='discussions-op-text'>

<sup>Originally posted by **inezpereira** July 31, 2023</sup>
Hello everyone!

I have a Logistic Regression model which uses Elastic Net regularization (specifically, an [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) with a log loss and an elasticnet penalty). I am using this model because I have a dataset with many relatively many features, and I wanted Elastic Net to do feature selection for me.
I have fitted my model and now calculated the associated SHAP values. I noticed however that, often, even if a feature weight has been set to zero (thanks to the L1 regularization in Elastic Net), I still get a non-zero SHAP value.

My question is: 
Is this a violation of missingness? 
Or is it indeed possible to have non-zero SHAP values even if the weight calculated for this feature is zero?

I've combed several talks by @slundberg, and had a look at [the original paper](https://papers.nips.cc/paper_files/paper/2017/file/8a20a8621978632d76c43dfd28b67767-Paper.pdf), but I am relatively new to SHAP, so your help would be very much appreciated :)

Below this code snippet I am using:
```
X_train_summary = shap.kmeans(X_train, 10)  
explainer = shap.KernelExplainer(model.predict_proba, X_train_summary, feature_names=model.feature_names_in_)
# Calculates the SHAP values - It takes some time
shap_values = explainer.shap_values(X_test)

```
Thanks!</div>

**Comment by znacer** [comment]
Hi,

I did not manage to reproduce this issue. I tried to build a model that ends with some null weights. The SHAP values are also null.
Here is the model I used :  
```py
import numpy as np
import shap
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier

# Seed for reproducibility
np.random.seed(0)

# Generate some data
N = 1000
# X_1 is biased towards class 0 or 1, X_2 is just noise
X = np.zeros((N, 3))
X[:, 0] = np.where(np.random.rand(N) > 0.5, np.random.rand(N)+1, np.random.rand(N)-1)
X[:, 1] = np.random.rand(N)
X[:, 2] = np.random.rand(N)
y = np.where(X[:, 0] > 0, 1, 0)

# Split data into 70% train and 30% test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

# Initialize and train classifier
model = SGDClassifier(loss="log_loss", penalty="l1", alpha=0.1)
model.fit(X_train, y_train)

# Print the feature weights
print(model.coef_)
# Output : [[2.13453436 0.         0.        ]]
#%%
X_train_summary = shap.kmeans(X_train, 10)  
explainer = shap.KernelExplainer(model.predict_proba, X_train_summary)
# Calculates the SHAP values - It takes some time
shap_values = explainer.shap_values(X_test)
print(shap_values[0][0])
# expected output: [-0.4242038  0.         0.       ]
```

**Comment by dsgibbons** [comment]
> Below this code snippet I am using:
> 
> ```
> X_train_summary = shap.kmeans(X_train, 10)  
> explainer = shap.KernelExplainer(model.predict_proba, X_train_summary, feature_names=model.feature_names_in_)
> # Calculates the SHAP values - It takes some time
> shap_values = explainer.shap_values(X_test)
> ```

@inezpereira this is not a reproducible example. Please provide a complete code block with imports and data that we can run.

**Comment by inezpereira** [comment]
Thank you for getting back to me on this! I am getting these results with a dataset that I cannot share.
I will try and reproduce the issue myself with a simple dataset and get back to you.

**Comment by inezpereira** [comment]
This was discussed in a separate thread and seems to theoretically be possible. See [here](https://github.com/shap/shap/discussions/3160#discussioncomment-6822813). I am closing this issue, as there is no bug.

**Comment by inezpereira** [comment]
This was discussed in a separate thread and seems to theoretically be possible. See [here](https://github.com/shap/shap/discussions/3160#discussioncomment-6822813). I am closing this issue, as there is no bug.
